caas/kubernetes/clientconfig/plugins.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package clientconfig
+
+import (
+	"github.com/juju/errors"
+
+	core "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp" // load gcp auth plugin.
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+)
+
+const (
+	adminNameSpace             = "kube-system"
+	clusterRoleName            = "cluster-admin"
+	jujuServiceAccountName     = "juju-service-account"
+	jujuClusterRoleBindingName = "juju-cluster-role-binding"
+)
+
+// To regenerate the mocks for the kubernetes Client used by this package,
+// mockgen -package mocks -destination mocks/rbacv1_mock.go k8s.io/client-go/kubernetes/typed/rbac/v1 RbacV1Interface,ClusterRoleBindingInterface,ClusterRoleInterface
+// mockgen -package mocks -destination mocks/serviceaccount_mock.go k8s.io/client-go/kubernetes/typed/core/v1 ServiceAccountInterface
+
+func ensureJujuAdminServiceAccount(
+	clientset kubernetes.Interface,
+	config *clientcmdapi.Config,
+	contextName string,
+) (*clientcmdapi.Config, error) {
+
+	// ensure admin cluster role.
+	clusterRole, err := ensureClusterRole(clientset, clusterRoleName, adminNameSpace)
+	if err != nil {
+		return nil, errors.Annotatef(
+			err, "ensuring cluster role %q in namespace %q", clusterRoleName, adminNameSpace)
+	}
+
+	// create juju admin service account.
+	sa, err := ensureServiceAccount(clientset, jujuServiceAccountName, adminNameSpace)
+	if err != nil {
+		return nil, errors.Annotatef(
+			err, "ensuring service account %q in namespace %q", jujuServiceAccountName, adminNameSpace)
+	}
+
+	// ensure role binding for juju admin service account with admin cluster role.
+	_, err = ensureClusterRoleBinding(clientset, jujuClusterRoleBindingName, sa, clusterRole)
+	if err != nil {
+		return nil, errors.Annotatef(err, "ensuring cluster role binding %q", jujuClusterRoleBindingName)
+	}
+
+	// refresh service account to get the secret/token after cluster role binding created.
+	sa, err = ensureServiceAccount(clientset, jujuServiceAccountName, adminNameSpace)
```

Comment:
let's call a straight Get() API here as it should exist and it's an error if it doesn't